### PR TITLE
Remove created directories on ninja uninstall

### PR DESCRIPTION
**Pull request by whot**
If a directory was created during ninja install, remove it again during ninja uninstall

https://github.com/mesonbuild/meson/issues/2032

### CODE REVIEW

File: mesonbuild/scripts/meson_install.py [hunk 1]
```diff
@@ -20,6 +20,35 @@
...
+        # store the directories in creation order, with the parent directory
+        # before the child directories. Future calls of makedir() will not
+        # create the parent directories, so the last element in the list is
+        # the last one to be created. That is the first one to be removed on
+        # __exit__
+        dirs.reverse()
+        self.dirs += dirs
+
+    def __enter__(self):
+        return self
```
Comment (ignatenkobrain) [inline mesonbuild/scripts/meson_install.py]: __[flake8]__
* [W291] trailing whitespace

File: mesonbuild/scripts/meson_install.py [hunk 2]
```diff
@@ -121,23 +150,24 @@ def do_install(datafilename):
...
     d.fullprefix = destdir_join(d.destdir, d.prefix)
 
-    install_subdirs(d) # Must be first, because it needs to delete the old subtree.
-    install_targets(d)
-    install_headers(d)
-    install_man(d)
-    install_data(d)
-    run_install_script(d)
+    d.dirmaker = DirMaker()
+    with d.dirmaker as dm:
```
Comment (ignatenkobrain) [inline mesonbuild/scripts/meson_install.py]: __[flake8]__
* [F841] local variable 'dm' is assigned to but never used

File: mesonbuild/scripts/meson_install.py [hunk 3]
```diff
@@ -20,6 +20,35 @@
...
 install_log_file = None
 
+class DirMaker:
+    def __init__(self):
+        self.dirs = []
+
+    def makedirs(self, path, exist_ok=False):
+        dirname = os.path.normpath(path)
+        dirs = []
+        while dirname != os.path.dirname(dirname):
```
Comment (ignatenkobrain) [inline mesonbuild/scripts/meson_install.py]: isn't this sufficient to try/except in os.mkdirs() without exist_ok?

File: mesonbuild/scripts/uninstall.py [hunk 4]
```diff
@@ -24,7 +24,10 @@ def do_uninstall(log):
             continue
         fname = line.strip()
         try:
-            os.unlink(fname)
+            if os.path.isdir(fname):
```
Comment (jpakkane) [inline mesonbuild/scripts/uninstall.py]: This returns true if fname is a symlink that points to a dir. It should probably have `follow_symlinks=False`.

File: mesonbuild/scripts/meson_install.py (hunk 3 above)
Comment (whot) [inline mesonbuild/scripts/meson_install.py]: sorry, I'm having trouble parsing this, can you rephrase please?

**Comment by whot** [comment]
squashed in the link check requested above, rebased on top of master and force-pushed.
